thetis/inversion_tools.py

```diff
@@ -0,0 +1,354 @@
+            f'{observation_data_dir}/'
+            f'diagnostic_timeseries_{s}_{variable}.hdf5' for s in
+            station_names
+        ]
+        # arrays of time stamps and values for each station
+        observation_time = []
+        observation_values = []
+        observation_coords = []
+        for f in file_list:
+            with h5py.File(f) as h5file:
+                t = h5file['time'][:].flatten()
+                v = h5file[variable][:].flatten()
+                x = h5file.attrs['x']
+                y = h5file.attrs['y']
+                observation_coords.append((x, y))
+                observation_time.append(t)
+                observation_values.append(v)
+        # list of station coordinates
+        observation_x, observation_y = numpy.array(observation_coords).T
+        self.register_observation_data(
+            station_names, variable, observation_time,
+            observation_values, observation_x, observation_y
+        )
+        self.construct_evaluator()
+
+    def construct_evaluator(self):
+        # Create 0D mesh for station evaluation
+        xy = numpy.array((self.observation_x, self.observation_y)).T
+        mesh0d = VertexOnlyMesh(self.mesh, xy)
+        self.fs_points_0d = FunctionSpace(mesh0d, 'DG', 0)
+        self.obs_values_0d = Function(self.fs_points_0d, name='observations')
+        self.mod_values_0d = Function(self.fs_points_0d, name='model values')
+
+        # Construct timeseries interpolator
+        self.station_interpolators = []
+        self.local_station_index = []
+        for i in range(self.fs_points_0d.dof_dset.size):
+            # loop over local DOFs and match coordinates to observations
+            # NOTE this must be done manually as VertexOnlyMesh reorders points
+            x_mesh, y_mesh = mesh0d.coordinates.dat.data[i, :]
+            xy_diff = xy - numpy.array([x_mesh, y_mesh])
+            xy_dist = numpy.sqrt(xy_diff[:, 0]**2 + xy_diff[:, 1]**2)
+            j = numpy.argmin(xy_dist)
+            self.local_station_index.append(j)
+
+            x, y = xy[j, :]
+            t = self.observation_time[j]
+            v = self.observation_values[j]
+            x_mesh, y_mesh = mesh0d.coordinates.dat.data[i, :]
+
+            msg = 'bad station location ' \
+                f'{j} {i} {x} {x_mesh} {y} {y_mesh} {x-x_mesh} {y-y_mesh}'
+            assert numpy.allclose([x, y], [x_mesh, y_mesh]), msg
+            # create temporal interpolator
+            ip = interp1d(t, v)
+            self.station_interpolators.append(ip)
+
+        # expressions for cost function
+        self.misfit_expr = self.obs_values_0d - self.mod_values_0d
+        self.initialzed = True
+
+    def eval_observation_at_time(self, t):
+        return [float(ip(t)) for ip in self.station_interpolators]
+
+    def eval_cost_function(self, t):
+        """
+        Evaluate the cost function.
+
+        Should be called at every export of the forward model.
+        """
```

Comment:
Worth mentioning here or somewhere else the cost function that we are implicitly assuming (sum of squares).